src/webgpu/api/operation/memory_sync/buffer/buffer_sync_test.ts

```diff
@@ -1,40 +1,276 @@
+          break;
+      }
+      return true;
+    };
+    return checkImpl(ops[0], ops[1]) && checkImpl(ops[1], ops[0]);
+  }
+  return true;
+}
+
+const kDummyVertexShader = `
+@stage(vertex) fn vert_main() -> @builtin(position) vec4<f32> {
+  return vec4<f32>(0.5, 0.5, 0.0, 1.0);
+}
+`;
 
 // Note: If it would be useful to have any of these helpers be separate from the fixture,
 // they can be refactored into standalone functions.
 export class BufferSyncTest extends GPUTest {
-  // Create a buffer, and initialize it to a specified value for all elements.
+  // Vertex and index buffers used in read render pass
+  vertexBuffer?: GPUBuffer;
+  indexBuffer?: GPUBuffer;
+
+  // Temp buffer and texture with values for buffer/texture copy write op
+  // There can be at most 2 write op
+  tmpValueBuffers: (GPUBuffer | undefined)[] = [undefined, undefined];
+  tmpValueTextures: (GPUTexture | undefined)[] = [undefined, undefined];
+
+  // Create extra buffers/textures needed by write operation
+  async createBuffersAndTexturesForWriteOp(writeOp: WriteOp, slot: number, value: number) {
+    switch (writeOp) {
+      case 'b2b-copy':
+        this.tmpValueBuffers[slot] = await this.createBufferWithValue(value);
+        break;
+      case 't2b-copy':
+        this.tmpValueTextures[slot] = await this.createTextureWithValue(value);
+        break;
+      default:
+        break;
+    }
+  }
+
+  // Create extra buffers/textures needed by read operation
+  async createBuffersAndTexturesForReadOp(readOp: ReadOp) {
+    switch (readOp) {
+      case 'input-index':
+        this.vertexBuffer = await this.createBufferWithValues([0, 1]);
+        break;
+      case 'input-indirect':
+        this.vertexBuffer = await this.createBufferWithValues([1]);
+        break;
+      case 'input-indirect-index':
+        this.vertexBuffer = await this.createBufferWithValues([1]);
+        this.indexBuffer = await this.createBufferWithValues([0]);
+        break;
+      default:
+        break;
+    }
+
+    let srcBuffer: GPUBuffer;
+    switch (readOp) {
+      case 'input-indirect':
+        // vertexCount = {0, 1}
+        // instanceCount = 1
+        // firstVertex = 0
+        // firstInstance = 0
+        srcBuffer = await this.createBufferWithValues([0, 1, 0, 0]);
+        break;
+      case 'input-indirect-index':
+        // indexCount = {0, 1}
+        // instanceCount = 1
+        // firstIndex = 0
+        // baseVertex = 0
+        // firstInstance = 0
+        srcBuffer = await this.createBufferWithValues([0, 1, 0, 0, 0]);
+        break;
+      case 'input-indirect-dispatch':
+        // workgroupCountX = {0, 1}
+        // workgroupCountY = 1
+        // workgroupCountZ = 1
+        srcBuffer = await this.createBufferWithValues([0, 1, 1]);
+        break;
+      default:
+        srcBuffer = await this.createBufferWithValue(0);
+        break;
+    }
+
+    const dstBuffer = await this.createBufferWithValue(0);
```

Comment:
would be best to not initialize to 0, so we don't have multiple writes going on which might affect the test